File: Android/ktv/src/main/java/com/tencent/cloud/uikit/livekit/songpicker/service/MusicService.kt
package com.tencent.cloud.uikit.livekit.songpicker.service;

import com.tencent.cloud.uikit.livekit.songpicker.model.Song
import com.tencent.cloud.uikit.livekit.songpicker.model.SongTag

interface MusicService {

    fun addObserver(observer: Observer)

    fun destroyService()

    fun getMusicTagList(callback: ValueCallback<List<SongTag>>?)

    fun getMusicsByTagId(
        tagId: String,
        scrollToken: String,
        callback: ValueCallback<List<Song>>?
    )

    fun addMusicToPlaylist(musicInfo: Song, callback: AddMusicCallback?)

    fun deleteMusicFromPlaylist(musicInfo: Song, callback: ActionCallback?)

    fun clearPlaylistByUserId(userId: String, callback: ActionCallback?)

    fun topMusic(musicInfo: Song, callback: ActionCallback?)

    fun switchMusicFromPlaylist(musicInfo: Song, callback: ActionCallback?)

    fun completePlaying(musicInfo: Song, callback: ActionCallback?)

    fun getMusicsByKeywords(
        scrollToken: String,
        pageSize: Int,
        keyWords: String,
        callback: ValueCallback<MusicInfoPage>?
    )

    fun getPlaylist(callback: ValueCallback<List<Song>>?)

    interface ActionCallback {
        fun onSuccess()
        fun onError(errorCode: Int, errorMessage: String)
    }

    interface ValueCallback<T> {
        fun onSuccess(value: T)
        fun onError(errorCode: Int, errorMessage: String)
    }

    interface AddMusicCallback {
        fun onStart(musicInfo: Song)
        fun onProgress(musicInfo: Song, progress: Float)
        fun onFinish(musicInfo: Song, errorCode: Int, errorMessage: String)
    }

    interface Observer {
        fun onMusicListChanged(musicInfoList: MutableList<Song>)
    }

    class MusicInfoPage(var musicList: List<Song> = emptyList(), var scrollToken: String = "")
}

File: iOS/TUILiveKit/Sources/KTV/SongPickerPanel/State/MusicService.swift
//
//  KaraokeMusicService.swift
//  TUIKaraoke
//
//  Created by gg on 2021/7/8.
//  Copyright © 2022 Tencent. All rights reserved.

import Foundation

public typealias KaraokeCallback = (_ code: Int, _ message: String) -> ()


public typealias MusicListCallback = (_ errorCode: NSInteger, _ errorMessage: String, _ list: [Song], _ scrollToken: String) -> Void

public typealias MusicTagListCallback = (_ errorCode: NSInteger, _ errorMessage: String, _ list: [MusicTagModel]) -> Void

public typealias MusicSelectedListCallback = (_ errorCode: NSInteger, _ errorMessage: String, _ list: [QueuedSong]) -> Void

public typealias MusicStartCallback = (_ musicInfo: QueuedSong) -> Void

public typealias MusicProgressCallback = (_ musicInfo: QueuedSong?, _ progress: Float) -> Void

public typealias MusicFinishCallback = (_ musicInfo: QueuedSong, _ errorCode: Int32, _ msg: String) -> Void

public typealias KaraokeAddMusicCallback = (start: MusicStartCallback, progress: MusicProgressCallback, finish: MusicFinishCallback)

public protocol KaraokeMusicService: AnyObject {
    
    func addObserver(_ observer: MusicServiceObserver)
    
    func destroyService()
    
    func setRoomId(_ roomId: String)
    
    func getMusicTagList(callback: @escaping MusicTagListCallback)
    
    func getMusicsByTagId(tagId: String, scrollToken: String, callback: @escaping MusicListCallback)
    
    func getMusicsByKeywords(keyWord: String, limit: Int, scrollToken: String, callback: @escaping MusicListCallback)
    
    func getPlaylist(_ callback: @escaping MusicSelectedListCallback)
    
    func addMusicToPlaylist(musicInfo: QueuedSong, callBack: KaraokeAddMusicCallback?)
    
    func deleteMusicFromPlaylist(musicInfo: QueuedSong, callback: @escaping KaraokeCallback)
    
    func clearPlaylistByUserId(userID: String, callback: @escaping KaraokeCallback)
    
    func topMusic(musicInfo: QueuedSong, callback: @escaping KaraokeCallback)
    
    func switchMusicFromPlaylist(musicInfo: QueuedSong, callback: @escaping KaraokeCallback)
    
    func completePlaying(musicInfo: QueuedSong, callback: @escaping KaraokeCallback)
    
    
    // MARK: - score
    func prepareMusicScore(musicInfo: QueuedSong)
    
    func processMusicScore(buffer: UnsafeMutablePointer<CChar>, length: Int32, timeStamp: Double)
 
    func finishMusicScore()
    
}
